File: src/labeling.ts
import { LabelerServer } from "@skyware/labeler";
import { FORTUNES, getDailyFortune } from "./fortune.js";
import { getJstTime } from "./utils.js";
// db import removed to keep this file pure/testable

/**
 * 指定された運勢以外の全ての運勢リスト（Negate対象）を返します。
 * @param currentFortune 現在の運勢
 * @returns 打ち消すべき運勢ラベルのリスト
 */
export function calculateNegateList(currentFortune: string): string[] {
    return FORTUNES.map((f) => f.val).filter((v) => v !== currentFortune);
}

/**
 * ユーザーに対して特定の運勢ラベルを付与し、それ以外の運勢ラベルを全て打ち消します (Negate)。
 * ギミック等の強制上書き用です。
 * @param did 対象ユーザーのDID
 * @param fortune 付与する運勢
 * @param labeler LabelerServerのインスタンス
 */
export async function overwriteFortune(did: string, fortune: string, labeler: LabelerServer) {
    const negate = calculateNegateList(fortune);
    try {
        await labeler.createLabels(
            { uri: did },
            {
                create: [fortune],
                negate: negate,
            }
        );
    } catch (e) {
        console.error(`Error applying fortune ${fortune} to ${did}:`, e);
    }
}

/**
 * ユーザーに対して日替わりの運勢ラベルを付与し、それ以外の運勢ラベルを全て打ち消します (Negate)。
 * @param did 対象ユーザーのDID
 * @param labeler LabelerServerのインスタンス
 */
export async function processUser(did: string, labeler: LabelerServer, handle?: string) {
    const fortune = getDailyFortune(did);
    const now = getJstTime();
    const identifier = handle ? `${handle} (${did})` : did;
    console.log(`[${now}] Processing ${identifier}, fortune: ${fortune}`);

    const negate = calculateNegateList(fortune);

    try {
        await labeler.createLabels(
            { uri: did },
            {
                create: [fortune],
                negate: negate,
            }
        );
    } catch (e) {
        console.error(`Error processing user ${did}:`, e);
    }
}

/**
 * ユーザーから全ての運勢ラベルを剥奪し (Opt-out)、ローカルDBからも削除します。
 * @param did 対象ユーザーのDID
 * @param labeler LabelerServerのインスタンス
 * @param db Databaseインスタンス (Dependency Injection)
 */
export async function negateUser(did: string, labeler: LabelerServer, db: any) {
    const now = getJstTime();
    console.log(`[${now}] Cleanup: Removing labels from ${did}`);
    const allFortunes = FORTUNES.map((f) => f.val);
    try {
        await labeler.createLabels({ uri: did }, { negate: allFortunes });
        db.prepare("DELETE FROM labels WHERE uri = ?").run(did);
    } catch (e) {
        console.error(`Failed to cleanup ${did}:`, e);
    }
}

File: src/labeling.rs
use crate::db::{DbPool, upsert_label as db_upsert, delete_label as db_delete};
use crate::fortune::{get_daily_fortune, FORTUNES};
use crate::crypto::sign_label;
use atrium_crypto::keypair::Secp256k1Keypair;
use atrium_api::com::atproto::label::defs::{Label, LabelData};
use atrium_api::types::string::{Datetime, Did};
use chrono::{Utc, SubsecRound};
use anyhow::Result;
use tokio::sync::broadcast;

pub async fn process_user(
    did: &str,
    handle: Option<&str>,
    pool: &DbPool,
    keypair: &Secp256k1Keypair,
    labeler_did: &str,
    tx: &broadcast::Sender<(i64, Vec<Label>)>
) -> Result<()> {
    let fortune_val = get_daily_fortune(did);
    println!("Processing {} ({:?}), fortune: {}", did, handle, fortune_val);

    let negate_list: Vec<&str> = FORTUNES.iter()
        .map(|f| f.val)
        .filter(|&v| v != fortune_val)
        .collect();

    upsert_label(did, fortune_val, false, labeler_did, pool, keypair, tx).await?;

    for neg_val in negate_list {
        upsert_label(did, neg_val, true, labeler_did, pool, keypair, tx).await?;
    }

    Ok(())
}

pub async fn overwrite_fortune(
    did: &str,
    fortune_val: &str,
    pool: &DbPool,
    keypair: &Secp256k1Keypair,
    labeler_did: &str,
    tx: &broadcast::Sender<(i64, Vec<Label>)>
) -> Result<()> {
    let negate_list: Vec<&str> = FORTUNES.iter()
        .map(|f| f.val)
        .filter(|&v| v != fortune_val)
        .collect();

    upsert_label(did, fortune_val, false, labeler_did, pool, keypair, tx).await?;
    for neg_val in negate_list {
        upsert_label(did, neg_val, true, labeler_did, pool, keypair, tx).await?;
    }
    Ok(())
}

pub async fn negate_user(
    did: &str,
    pool: &DbPool,
    _keypair: &Secp256k1Keypair,
    _labeler_did: &str,
    _tx: &broadcast::Sender<(i64, Vec<Label>)>
) -> Result<()> {
    db_delete(pool, did).await?;
    Ok(())
}

async fn upsert_label(
    uri: &str,
    val: &str,
    neg: bool,
    src: &str,
    pool: &DbPool,
    keypair: &Secp256k1Keypair,
    tx: &broadcast::Sender<(i64, Vec<Label>)>
) -> Result<()> {
    let now = Utc::now().with_timezone(&chrono::FixedOffset::east_opt(0).unwrap()).round_subsecs(3);
    let cts = Datetime::new(now);

    let mut label_data = LabelData {
        cid: None,
        cts: cts.clone(),
        exp: None,
        neg: if neg { Some(true) } else { None },
        sig: None,
        src: Did::new(src.to_string()).expect("Invalid DID"), // Ensure config DID is valid
        uri: uri.to_string(),
        val: val.to_string(),
        ver: None,
    };

    sign_label(&mut label_data, keypair)?;

    let rowid = db_upsert(pool, uri, val, &cts.as_ref().to_rfc3339(), neg, src).await?;

    // Create Label struct for broadcast
    let label = Label {
        data: label_data.clone(),
        extra_data: ipld_core::ipld::Ipld::Null,
    };

    // Broadcast
    match tx.send((rowid, vec![label])) {
        Ok(count) => println!("Broadcaster: To {} listeners. Seq={}, Val={}", count, rowid, val),
        Err(_) => println!("Broadcaster: No listeners active. Seq={}, Val={}", rowid, val),
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::{init_db, get_labels};

    #[tokio::test]
    async fn test_process_user_logic() -> Result<()> {
        let pool = init_db(":memory:").await?;
        use rand::rngs::OsRng;
        let mut rng = OsRng;
        let keypair = Secp256k1Keypair::create(&mut rng);
        let labeler_did = "did:plc:labeler";
        let target_did = "did:plc:target";
        let (tx, _rx) = broadcast::channel(100);

        process_user(target_did, None, &pool, &keypair, labeler_did, &tx).await?;

        let labels = get_labels(&pool, target_did, None, None).await?;
        assert!(!labels.is_empty(), "Labels should be created");

        let positives: Vec<_> = labels.iter().filter(|l| l.neg == 0).collect();
        let negatives: Vec<_> = labels.iter().filter(|l| l.neg == 1).collect();

        assert_eq!(positives.len(), 1, "Should have exactly 1 positive label");
        assert_eq!(negatives.len(), 6, "Should have 6 negative labels");
        assert_eq!(labels.len(), 7, "Total labels should be 7");

        println!("Granted fortune: {}", positives[0].val);

        Ok(())
    }
}
